api/server/api.cc

```diff
@@ -283,6 +283,42 @@ void API::sg_update(const app::Sg &sg, const app::Rule &rule) {
     }
 }
 
+void API::sg_update_rule_members(const app::Sg &modified_sg) {
+    std::map<std::string, app::Nic>::iterator it;
+    std::vector<std::string>::iterator sg_it;
+    std::map<std::size_t, app::Rule>::iterator rule_it;
+    std::vector<std::string> updated_sgs;
+    auto found = false;
+    for (it = app::model.nics.begin(); it != app::model.nics.end(); it++) {
+        app::Nic &nic = it->second;
+        for (sg_it = nic.security_groups.begin();
+            sg_it != nic.security_groups.end();
+            sg_it++) {
+            auto sg = app::model.security_groups.find(*sg_it);
+            if (sg == app::model.security_groups.end())
+                continue;
+            for (rule_it = sg->second.rules.begin();
+                 rule_it != sg->second.rules.end();
```

Comment:
same here, end should not be call at each iteration of your for, except if you change your container during the for.